### PR TITLE
enhance consensus key rotation support

### PR DESCRIPTION
**Pull request by zjma**
### Consensus key rotation is not well-supported today

Today, a validator only uses a single consensus key, and the only way to update the consensus key is to edit the node config and restart the node...

- when a validator updates consensus pk on chain in epoch `e` then updates local consensus sk in epoch `e+1`, incorrect augmented keys for randomness has been persisted at the beginning of epoch `e+1`, and node will crash on restart (until epoch `e+2` when the incorrect data is cleared).
- when a validator updates consensus pk on chain in epoch `e` and updates local consensus sk also in epoch `e`, the node won't be able to participate in consensus or randomness until epoch `e+1`.

### Proposed user flow
- Stop the node
- In additional to the old identity, specify the new key in node config.
- Restart the node (and the validator should still function with the old key)
- Run a txn to update pk for next epoch
- When the next epoch arrives, the validator should automatically switch to the new key without any operation/downtime.

### The code changes

- Add a node config entry: `NodeConfig::ConsensusConfig::SafetyRulesConfig::InitialSafetyRulesConfig::FromFile::overriding_identity_paths`, which is to hold new validator identitie(s).
- Each time the secure storage is (re-)initialized, (typically every time the node restarts), for every validator identity given by the entry above, its private key is saved into the secure storage under key `consensus_<PK_HEX>` where `PK_HEX` is the public key hexlified.
- At the beginning of a new epoch, all the consensus key-dependent components now find the consensus sk in the secure storage by `consensus_<X>` where X is its on-chain validator pk hexlified. If they don't exist, fall back to the value under key `consensus` (set by the existing implementation).
- A smoke test `consensus_key_rotation` to capture the user flow.

## Type of Change
- [x] New feature
- [x] Bug fix
- [x] Refactoring

## Which Components or Systems Does This Change Impact?
- [x] Validator Node

## How Has This Been Tested?

A smoke test case.

## Key Areas to Review

The entire PR.

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: crates/aptos-dkg/src/pvss/encryption_dlog.rs [hunk 1]
```diff
@@ -211,7 +211,122 @@ macro_rules! encryption_dlog_keys_impl {
 
 pub mod g1 {
     // PPs not needed, for now.
-    encryption_dlog_keys_impl!(G1_PROJ_NUM_BYTES, g1_proj_from_bytes, G1Projective);
```
Comment (zjma) [inline crates/aptos-dkg/src/pvss/encryption_dlog.rs:214]: revert

File: crates/aptos-genesis/src/builder.rs [hunk 2]
```diff
@@ -166,7 +166,11 @@ impl ValidatorNodeConfig {
         // Init safety rules
         let validator_identity_file = self.dir.join(VALIDATOR_IDENTITY);
         config.consensus.safety_rules.initial_safety_rules_config =
-            InitialSafetyRulesConfig::from_file(validator_identity_file, waypoint_config.clone());
```
Comment (zjma) [inline crates/aptos-genesis/src/builder.rs:169]: revert

File: config/src/config/safety_rules_config.rs [hunk 3]
```diff
@@ -160,6 +167,37 @@ impl InitialSafetyRulesConfig {
...
+                }
+                Ok(blobs)
+            },
+            InitialSafetyRulesConfig::None => {
+                bail!("loading overriding identity blobs failed with missing initial safety rules config")
+            },
+        }
+    }
+
+    pub fn overriding_identity_blob_paths_mut(&mut self) -> &mut Vec<PathBuf> {
```
Comment (JoshLind) [inline config/src/config/safety_rules_config.rs:190]: nit: Maybe make this test-only? It seems like it's only used in the smoke tests? Or better yet, don't expose this here if possible?

File: config/src/config/safety_rules_config.rs [hunk 4]
```diff
@@ -123,15 +123,22 @@ impl ConfigSanitizer for SafetyRulesConfig {
 pub enum InitialSafetyRulesConfig {
     FromFile {
         identity_blob_path: PathBuf,
+        #[serde(skip_serializing_if = "Vec::is_empty", default)]
+        overriding_identity_paths: Vec<PathBuf>,
```
Comment (JoshLind) [inline config/src/config/safety_rules_config.rs:127]: Does this need to be a vector? Will 1 path not suffice (if the operator wants to rotate again, they need to move the new key to the old key location, and then add a single override). Would that work?

**Review by JoshLind — COMMENTED** [review]
Code seems reasonable to me.

Only big question is where are we planning to document this flow? It seems pretty precise to get right 😄

File: consensus/safety-rules/src/persistent_safety_storage.rs [hunk 5]
```diff
@@ -66,6 +66,10 @@ impl PersistentSafetyStorage {
         author: Author,
         consensus_private_key: bls12381::PrivateKey,
     ) -> Result<(), Error> {
+        warn!(
+            "0704 - writing consensus key 0x{} into secure storage",
```
[the remark this answers is not in the document]
Comment (zjma) [reply consensus/safety-rules/src/persistent_safety_storage.rs:70]: debugging stuff 🤦  will delete

File: consensus/safety-rules/src/persistent_safety_storage.rs [hunk 6]
```diff
@@ -101,10 +105,33 @@ impl PersistentSafetyStorage {
...
+                info!("0704 - sk_0");
+                sk_0
+            },
+            (Err(_), Ok(sk_1)) => {
+                info!("0704 - sk_1");
+                sk_1
+            },
+            (Err(_), Err(_)) => {
+                info!("0704 - err");
+                return Err(Error::ValidatorKeyNotFound("0704 - not found!".to_string()));
```
[the remark this answers is not in the document]
Comment (zjma) [reply consensus/safety-rules/src/persistent_safety_storage.rs:129]: debugging stuff 🤦  will delete

File: config/src/config/safety_rules_config.rs (hunk 4 above)
Comment (zjma) [reply config/src/config/safety_rules_config.rs:127]: In theory 2 slots (1 existing, 1 new) is enough. It's just i found n slots can also be supported for free by just replacing `Option<PathBuf>` with `Vec<PathBuf>`...

**Comment by zjma** [comment]
> Code seems reasonable to me.
> 
> Only big question is where are we planning to document this flow? It seems pretty challenging to get right without detailed instructions 😄

@JoshLind i plan to update [this](https://aptos.dev/en/network/nodes/validator-node/connect-nodes/connect-to-aptos-network#4-update-on-chain-consensus-key) after this is merged.

File: consensus/safety-rules/src/safety_rules_manager.rs [hunk 7]
```diff
@@ -75,19 +75,32 @@ pub fn storage(config: &SafetyRulesConfig) -> PersistentSafetyStorage {
             panic!(
                 "Safety rules storage is not initialized, provide an initial safety rules config"
             )
+        };
+
+        // Ensuring all the overriding consensus keys are in the storage.
+        let timer = Instant::now();
+        for blob in config
```
Comment (zekun000) [inline consensus/safety-rules/src/safety_rules_manager.rs:82]: I don't think this path is executed on every restart, it's only executed when the storage is empty? i.e. line 49 returns err

File: consensus/src/pipeline/execution_client.rs [hunk 8]
```diff
@@ -215,10 +217,9 @@ impl ExecutionProxyClient {
                 let (rand_ready_block_tx, rand_ready_block_rx) = unbounded::<OrderedBlocks>();
 
                 let (reset_tx_to_rand_manager, reset_rand_manager_rx) = unbounded::<ResetRequest>();
-                let consensus_key =
-                    load_consensus_key_from_secure_storage(&self.consensus_config.safety_rules)
-                        .expect("Failed in loading consensus key for ExecutionProxyClient.");
-                let signer = Arc::new(ValidatorSigner::new(self.author, consensus_key));
+                let consensus_sk = maybe_consensus_key
+                    .expect("consensus key unavailable for ExecutionProxyClient");
```
Comment (zekun000) [inline consensus/src/pipeline/execution_client.rs:221]: given we just panic if the option doesn't exist, we should probably pass in Arc directly?

File: crates/aptos/src/test/mod.rs [hunk 9]
```diff
@@ -528,9 +528,10 @@ impl CliTestFramework {
         pool_index: Option<usize>,
         consensus_public_key: bls12381::PublicKey,
         proof_of_possession: bls12381::ProofOfPossession,
+        gas_options: Option<GasOptions>,
     ) -> CliTypedResult<TransactionSummary> {
         UpdateConsensusKey {
-            txn_options: self.transaction_options(operator_index, None),
+            txn_options: self.transaction_options(operator_index, gas_options),
```
Comment (zekun000) [inline crates/aptos/src/test/mod.rs:534]: how's this change related?

**Review by zekun000 — COMMENTED** [review]
looks reasonable, let's replace the debug logs with meaningful logs. we may want to support truncating old keys too but can do it later

File: crates/aptos/src/test/mod.rs (hunk 9 above)
Comment (zjma) [reply crates/aptos/src/test/mod.rs:534]: needed for a smoke test to be less flaky.

File: consensus/safety-rules/src/safety_rules_manager.rs (hunk 7 above)
Comment (zjma) [reply consensus/safety-rules/src/safety_rules_manager.rs:82]: i believe it is? L49 and L80 are at the same indentation level, and there's no `return` in block L49-78.

Comment (zekun000) [reply consensus/safety-rules/src/safety_rules_manager.rs:82]: this block of line 55-102 is an else branch of line 50? https://github.com/aptos-labs/aptos-core/blob/869faed30c3868aec7879fc30701d545b841d745/consensus/safety-rules/src/safety_rules_manager.rs#L50

Comment (zekun000) [reply consensus/safety-rules/src/safety_rules_manager.rs:82]: nvm I'm noob

File: consensus/src/pipeline/execution_client.rs (hunk 8 above)
Comment (zjma) [reply consensus/src/pipeline/execution_client.rs:221]: we unwrap only if the current node is in the validator set.
For non-validators i guess we can't unwrap outside...

File: config/src/config/safety_rules_config.rs (hunk 3 above)
Comment (zjma) [reply config/src/config/safety_rules_config.rs:191]: fixed.

NOTE that we can't reuse feature `testing` as crate `smoke-test` needs it disabled in addition to `overriding_identity_blob_paths_mut`.

Comment (JoshLind) [reply config/src/config/safety_rules_config.rs:191]: Aah, in that case, don't worry about it -- you can remove the `smoke-test` feature. I was just wondering if it was a quick change 😄 Seems like overkill for a new feature.

File: testsuite/smoke-test/src/consensus_key_rotation.rs [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+) -> anyhow::Result<()> {
+    let timer = Instant::now();
+    while timer.elapsed() < time_limit {
+        let c = get_on_chain_resource::<ConfigurationResource>(rest_cli).await;
+        if c.epoch() >= target_epoch {
+            return Ok(());
+        }
+        tokio::time::sleep(Duration::from_secs(1)).await;
+    }
+    bail!("");
```
Comment (JoshLind) [inline testsuite/smoke-test/src/consensus_key_rotation.rs:191]: nit: maybe put a real error message here? 😄

File: testsuite/smoke-test/src/consensus_key_rotation.rs [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+                .push(new_identity_path);
+            validator_override_config.save_config(config_path).unwrap();
+
+            info!("Restarting the node.");
+            validator.start().unwrap();
+            info!("Let it bake for 5 secs.");
+            tokio::time::sleep(Duration::from_secs(5)).await;
+            (operator_addr, new_pk, pop, operator_idx)
+        } else {
+            unreachable!()
```
[the remark this answers is not in the document]
Comment (JoshLind) [reply testsuite/smoke-test/src/consensus_key_rotation.rs:118]: Still think the if-let is superfluous 😄

File: consensus/safety-rules/src/safety_rules_manager.rs [hunk 12]
```diff
@@ -75,19 +75,32 @@ pub fn storage(config: &SafetyRulesConfig) -> PersistentSafetyStorage {
...
+                    },
+                    Err(e) => {
+                        warn!("Setting {storage_key} failed with internal store set error: {e}");
+                    },
+                }
+            }
         }
-    }
-}
+        info!("Overriding key work time: {:?}", timer.elapsed());
```
Comment (JoshLind) [inline consensus/safety-rules/src/safety_rules_manager.rs:100]: nit: do we think this is necessary to track/log?

**Review by JoshLind — COMMENTED** [review]
Looks reasonable to me (from the node config and smoke test side). But, I defer the final stamp to @zekun000 for the consensus and validator verifier changes (some of this code is very old 😄).

**Review by zekun000 — APPROVED** [review]

**Review by JoshLind — APPROVED** [review]

File: testsuite/smoke-test/src/consensus_key_rotation.rs (hunk 10 above)
Comment (zjma) [reply testsuite/smoke-test/src/consensus_key_rotation.rs:191]: oops, i forgot auto-merge is on...

**Pull request merged**